File: qomm_transport/order.py
"""Content-independent admission and ordering for one fixed market slot.

The payload cannot decide priority.  A KYB admission authority issues exactly
one opaque ticket to each legal entity before the slot's randomness exists.
After the deadline, an independently signed beacon permutes those tickets.  A
ticket holder receives a signed admission receipt, so omission is challengeable
without disclosing the request carried by the frame.

Every admitted entity is expected to send one frame in every slot, including a
cover frame when it has no request.  Consequently the public manifest has a
fixed number of entries and does not reveal how many requests were real.
"""

from __future__ import annotations

import hashlib
import hmac
import time
from dataclasses import dataclass, field
from typing import Callable, Iterable

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .wire import Frame

TICKET_DOMAIN = b"QOMM:ORDER:TICKET:v1"
RECEIPT_DOMAIN = b"QOMM:ORDER:RECEIPT:v1"
BEACON_DOMAIN = b"QOMM:ORDER:BEACON:v1"
MANIFEST_DOMAIN = b"QOMM:ORDER:MANIFEST:v1"
ORDER_DOMAIN = b"QOMM:ORDER:KEY:v1"
ZERO = b"\x00" * 32


def _u64(value: int) -> bytes:
    if not 0 <= value < 1 << 64:
        raise ValueError("integer is outside the unsigned 64-bit range")
    return value.to_bytes(8, "big")


def _digest(*parts: bytes) -> bytes:
    h = hashlib.sha256()
    for part in parts:
        h.update(len(part).to_bytes(4, "big"))
        h.update(part)
    return h.digest()


def _verify(key: Ed25519PublicKey, signature: bytes, body: bytes) -> bool:
    try:
        key.verify(signature, body)
        return True
    except InvalidSignature:
        return False


@dataclass(frozen=True)
class AdmissionTicket:
    slot: int
    ticket_id: bytes
    issued_at: int
    expires_at: int
    signature: bytes

    def unsigned(self) -> bytes:
        if len(self.ticket_id) != 32:
            raise ValueError("ticket_id must be 32 bytes")
        if self.expires_at <= self.issued_at:
            raise ValueError("ticket expiry must follow issuance")
        return (TICKET_DOMAIN + _u64(self.slot) + self.ticket_id
                + _u64(self.issued_at) + _u64(self.expires_at))

    @property
    def digest(self) -> bytes:
        return _digest(self.unsigned(), self.signature)

    def verify(self, authority: Ed25519PublicKey, now: int) -> bool:
        return (self.issued_at <= now <= self.expires_at
                and _verify(authority, self.signature, self.unsigned()))


class AdmissionAuthority:
    """Issues one unlinkable ticket per verified entity and slot.

    ``entity_id`` never appears in the ticket.  Its HMAC nullifier is retained
    only to enforce the one-entity/one-ticket rule.  Production deployments must
    persist ``_issued`` transactionally; :mod:`qomm_transport.key_management`
    provides the encrypted store used by the command-line service.
    """

    def __init__(self, signing_key: Ed25519PrivateKey, entity_key: bytes):
        if len(entity_key) < 32:
            raise ValueError("entity nullifier key must contain at least 32 bytes")
        self.signing_key = signing_key
        self.entity_key = entity_key
        self._issued: dict[tuple[int, bytes], bytes] = {}

    @property
    def verifying_key(self) -> Ed25519PublicKey:
        return self.signing_key.public_key()

    def _entity_nullifier(self, entity_id: bytes) -> bytes:
        return hmac.new(self.entity_key, b"QOMM:ENTITY:v1" + entity_id,
                        hashlib.sha256).digest()

    def issue(self, entity_id: bytes, slot: int, *, issued_at: int | None = None,
              lifetime: int = 300, ticket_id: bytes | None = None) -> AdmissionTicket:
        if not entity_id:
            raise ValueError("an empty legal-entity identifier is not admissible")
        if lifetime <= 0:
            raise ValueError("ticket lifetime must be positive")
        issued_at = int(time.time()) if issued_at is None else issued_at
        nullifier = self._entity_nullifier(entity_id)
        key = (slot, nullifier)
        if key in self._issued:
            raise ValueError("this legal entity already has a ticket for the slot")
        if ticket_id is None:
            import os
            ticket_id = os.urandom(32)
        unsigned = (TICKET_DOMAIN + _u64(slot) + ticket_id
                    + _u64(issued_at) + _u64(issued_at + lifetime))
        ticket = AdmissionTicket(slot, ticket_id, issued_at, issued_at + lifetime,
                                 self.signing_key.sign(unsigned))
        self._issued[key] = ticket.digest
        return ticket


@dataclass(frozen=True)
class RandomnessBeacon:
    round: int
    value: bytes
    signature: bytes

    def unsigned(self) -> bytes:
        if len(self.value) != 32:
            raise ValueError("beacon value must be 32 bytes")
        return BEACON_DOMAIN + _u64(self.round) + self.value

    def verify(self, key: Ed25519PublicKey) -> bool:
        return _verify(key, self.signature, self.unsigned())

    @classmethod
    def sign(cls, round_: int, value: bytes,
             key: Ed25519PrivateKey) -> "RandomnessBeacon":
        unsigned = BEACON_DOMAIN + _u64(round_) + value
        return cls(round_, value, key.sign(unsigned))


@dataclass(frozen=True)
class AdmissionReceipt:
    slot: int
    node: int
    ticket_digest: bytes
    frame_digest: bytes
    received_at_ns: int
    signature: bytes

    def unsigned(self) -> bytes:
        if len(self.ticket_digest) != 32 or len(self.frame_digest) != 32:
            raise ValueError("receipt digests must be 32 bytes")
        return (RECEIPT_DOMAIN + _u64(self.slot) + _u64(self.node)
                + self.ticket_digest + self.frame_digest + _u64(self.received_at_ns))

    def verify(self, sealer_key: Ed25519PublicKey) -> bool:
        return _verify(sealer_key, self.signature, self.unsigned())


@dataclass(frozen=True)
class BatchManifest:
    slot: int
    node: int
    beacon_round: int
    beacon_value: bytes
    ordered_ticket_digests: tuple[bytes, ...]
    ordered_frame_digests: tuple[bytes, ...]
    previous_digest: bytes
    signature: bytes

    def unsigned(self) -> bytes:
        if len(self.beacon_value) != 32 or len(self.previous_digest) != 32:
            raise ValueError("manifest digests must be 32 bytes")
        if len(self.ordered_ticket_digests) != len(self.ordered_frame_digests):
            raise ValueError("ticket and frame manifests have different lengths")
        body = (MANIFEST_DOMAIN + _u64(self.slot) + _u64(self.node)
                + _u64(self.beacon_round) + self.beacon_value
                + self.previous_digest + _u64(len(self.ordered_ticket_digests)))
        for ticket, frame in zip(self.ordered_ticket_digests,
                                 self.ordered_frame_digests, strict=True):
            if len(ticket) != 32 or len(frame) != 32:
                raise ValueError("manifest entry digests must be 32 bytes")
            body += ticket + frame
        return body

    @property
    def digest(self) -> bytes:
        return _digest(self.unsigned(), self.signature)

    def verify(self, sealer_key: Ed25519PublicKey) -> bool:
        return _verify(sealer_key, self.signature, self.unsigned())

    def includes(self, receipt: AdmissionReceipt) -> bool:
        return any(t == receipt.ticket_digest and f == receipt.frame_digest
                   for t, f in zip(self.ordered_ticket_digests,
                                   self.ordered_frame_digests, strict=True))


@dataclass
class FixedSlotSealer:
    """Collects one frame per pre-issued ticket and closes exactly once."""

    slot: int
    node: int
    deadline_ns: int
    tickets: tuple[AdmissionTicket, ...]
    authority_key: Ed25519PublicKey
    beacon_key: Ed25519PublicKey
    signing_key: Ed25519PrivateKey
    previous_digest: bytes = ZERO
    _frames: dict[bytes, Frame] = field(default_factory=dict, init=False)
    _receipts: dict[bytes, AdmissionReceipt] = field(default_factory=dict, init=False)
    _closed: bool = field(default=False, init=False)

    def __post_init__(self) -> None:
        if len(self.previous_digest) != 32:
            raise ValueError("previous manifest digest must be 32 bytes")
        if len({ticket.ticket_id for ticket in self.tickets}) != len(self.tickets):
            raise ValueError("the expected ticket list contains a duplicate")
        for ticket in self.tickets:
            if ticket.slot != self.slot:
                raise ValueError("a ticket belongs to another slot")

    @property
    def verifying_key(self) -> Ed25519PublicKey:
        return self.signing_key.public_key()

    def admit(self, ticket: AdmissionTicket, frame: Frame, *, now_ns: int) -> AdmissionReceipt:
        if self._closed:
            raise RuntimeError("the slot is already closed")
        if now_ns > self.deadline_ns:
            raise TimeoutError("the frame arrived after the sealed deadline")
        now = now_ns // 1_000_000_000
        if not ticket.verify(self.authority_key, now):
            raise ValueError("the admission ticket is invalid or expired")
        expected = {item.digest for item in self.tickets}
        if ticket.digest not in expected:
            raise ValueError("the ticket was not in the slot's precommitted population")
        if frame.slot != self.slot or frame.node != self.node:
            raise ValueError("the frame belongs to another slot or node")
        raw = frame.encode()
        frame_digest = hashlib.sha256(raw).digest()
        prior = self._frames.get(ticket.digest)
        if prior is not None:
            if prior.encode() != raw:
                raise ValueError("one ticket attempted to replace its admitted frame")
            return self._receipts[ticket.digest]
        unsigned = (RECEIPT_DOMAIN + _u64(self.slot) + _u64(self.node)
                    + ticket.digest + frame_digest + _u64(now_ns))
        receipt = AdmissionReceipt(self.slot, self.node, ticket.digest,
                                   frame_digest, now_ns,
                                   self.signing_key.sign(unsigned))
        self._frames[ticket.digest] = frame
        self._receipts[ticket.digest] = receipt
        return receipt

    def close(self, beacon: RandomnessBeacon, *, now_ns: int
              ) -> tuple[list[Frame], BatchManifest]:
        if self._closed:
            raise RuntimeError("the slot is already closed")
        if now_ns <= self.deadline_ns:
            raise TimeoutError("the batch cannot close before its deadline")
        if not beacon.verify(self.beacon_key):
            raise ValueError("the ordering beacon signature is invalid")
        if beacon.round <= self.slot:
            raise ValueError("ordering randomness must be generated after the slot")
        missing = [ticket.digest for ticket in self.tickets
                   if ticket.digest not in self._frames]
        if missing:
            raise RuntimeError(
                f"fixed population incomplete: {len(missing)} cover or request frame(s) missing")
        ordered = sorted(
            self.tickets,
            key=lambda ticket: hashlib.sha256(
                ORDER_DOMAIN + _u64(self.slot) + _u64(beacon.round)
                + beacon.value + ticket.ticket_id).digest(),
        )
        frames = [self._frames[ticket.digest] for ticket in ordered]
        ticket_digests = tuple(ticket.digest for ticket in ordered)
        frame_digests = tuple(hashlib.sha256(frame.encode()).digest() for frame in frames)
        unsigned = (MANIFEST_DOMAIN + _u64(self.slot) + _u64(self.node)
                    + _u64(beacon.round) + beacon.value + self.previous_digest
                    + _u64(len(ticket_digests))
                    + b"".join(t + f for t, f in zip(ticket_digests, frame_digests,
                                                       strict=True)))
        manifest = BatchManifest(
            self.slot, self.node, beacon.round, beacon.value,
            ticket_digests, frame_digests, self.previous_digest,
            self.signing_key.sign(unsigned),
        )
        self._closed = True
        return frames, manifest


def prove_omission(receipt: AdmissionReceipt, manifest: BatchManifest,
                   sealer_key: Ed25519PublicKey) -> bool:
    """A public, payload-free omission challenge.

    ``True`` means the sealer signed both an admission receipt and a closed
    manifest that omitted that exact ticket/frame pair.
    """

    return (receipt.slot == manifest.slot and receipt.node == manifest.node
            and receipt.verify(sealer_key) and manifest.verify(sealer_key)
            and not manifest.includes(receipt))

File: rust/qomm-transport/src/order.rs
//! Content-independent admission and ordering for one fixed market slot.

use crate::wire::Frame;
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use sha2::{Digest, Sha256};
use std::collections::{BTreeMap, BTreeSet};
use std::time::{SystemTime, UNIX_EPOCH};

const TICKET_DOMAIN: &[u8] = b"QOMM:ORDER:TICKET:v1";
const RECEIPT_DOMAIN: &[u8] = b"QOMM:ORDER:RECEIPT:v1";
const BEACON_DOMAIN: &[u8] = b"QOMM:ORDER:BEACON:v1";
const MANIFEST_DOMAIN: &[u8] = b"QOMM:ORDER:MANIFEST:v1";
const ORDER_DOMAIN: &[u8] = b"QOMM:ORDER:KEY:v1";
pub const ZERO: [u8; 32] = [0; 32];

fn digest(parts: &[&[u8]]) -> [u8; 32] {
    let mut hash = Sha256::new();
    for part in parts {
        hash.update((part.len() as u32).to_be_bytes());
        hash.update(part);
    }
    hash.finalize().into()
}

fn hmac(key: &[u8], body: &[u8]) -> [u8; 32] {
    let mut block = [0_u8; 64];
    if key.len() > 64 {
        block[..32].copy_from_slice(&Sha256::digest(key));
    } else {
        block[..key.len()].copy_from_slice(key);
    }
    let mut inner_pad = [0x36_u8; 64];
    let mut outer_pad = [0x5c_u8; 64];
    for index in 0..64 {
        inner_pad[index] ^= block[index];
        outer_pad[index] ^= block[index];
    }
    let inner = Sha256::new()
        .chain_update(inner_pad)
        .chain_update(body)
        .finalize();
    Sha256::new()
        .chain_update(outer_pad)
        .chain_update(inner)
        .finalize()
        .into()
}

#[derive(Clone, Debug)]
pub struct AdmissionTicket {
    pub slot: u64,
    pub ticket_id: [u8; 32],
    pub issued_at: u64,
    pub expires_at: u64,
    pub signature: Signature,
}

impl AdmissionTicket {
    pub fn unsigned(&self) -> Result<Vec<u8>, String> {
        if self.expires_at <= self.issued_at {
            return Err("ticket expiry must follow issuance".into());
        }
        let mut body = Vec::with_capacity(TICKET_DOMAIN.len() + 56);
        body.extend_from_slice(TICKET_DOMAIN);
        body.extend_from_slice(&self.slot.to_be_bytes());
        body.extend_from_slice(&self.ticket_id);
        body.extend_from_slice(&self.issued_at.to_be_bytes());
        body.extend_from_slice(&self.expires_at.to_be_bytes());
        Ok(body)
    }

    pub fn digest(&self) -> Result<[u8; 32], String> {
        Ok(digest(&[&self.unsigned()?, &self.signature.to_bytes()]))
    }

    pub fn verify(&self, authority: &VerifyingKey, now: u64) -> bool {
        self.issued_at <= now
            && now <= self.expires_at
            && self
                .unsigned()
                .is_ok_and(|body| authority.verify(&body, &self.signature).is_ok())
    }
}

pub struct AdmissionAuthority {
    signing_key: SigningKey,
    entity_key: Vec<u8>,
    issued: BTreeMap<(u64, [u8; 32]), [u8; 32]>,
}

impl AdmissionAuthority {
    pub fn new(signing_key: SigningKey, entity_key: Vec<u8>) -> Result<Self, String> {
        if entity_key.len() < 32 {
            return Err("entity nullifier key must contain at least 32 bytes".into());
        }
        Ok(Self {
            signing_key,
            entity_key,
            issued: BTreeMap::new(),
        })
    }

    pub fn verifying_key(&self) -> VerifyingKey {
        self.signing_key.verifying_key()
    }

    pub fn entity_nullifier(&self, entity_id: &[u8]) -> Result<[u8; 32], String> {
        if entity_id.is_empty() {
            return Err("an empty legal-entity identifier is not admissible".into());
        }
        let mut body = Vec::with_capacity(14 + entity_id.len());
        body.extend_from_slice(b"QOMM:ENTITY:v1");
        body.extend_from_slice(entity_id);
        Ok(hmac(&self.entity_key, &body))
    }

    pub fn issue(
        &mut self,
        entity_id: &[u8],
        slot: u64,
        issued_at: Option<u64>,
        lifetime: u64,
        ticket_id: Option<[u8; 32]>,
    ) -> Result<AdmissionTicket, String> {
        if lifetime == 0 {
            return Err("ticket lifetime must be positive".into());
        }
        let issued_at = issued_at.unwrap_or_else(|| {
            SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs()
        });
        let expires_at = issued_at
            .checked_add(lifetime)
            .ok_or_else(|| "ticket expiry overflow".to_string())?;
        let nullifier = self.entity_nullifier(entity_id)?;
        if self.issued.contains_key(&(slot, nullifier)) {
            return Err("this legal entity already has a ticket for the slot".into());
        }
        let ticket_id = ticket_id.unwrap_or_else(rand::random);
        let mut body = Vec::new();
        body.extend_from_slice(TICKET_DOMAIN);
        body.extend_from_slice(&slot.to_be_bytes());
        body.extend_from_slice(&ticket_id);
        body.extend_from_slice(&issued_at.to_be_bytes());
        body.extend_from_slice(&expires_at.to_be_bytes());
        let ticket = AdmissionTicket {
            slot,
            ticket_id,
            issued_at,
            expires_at,
            signature: self.signing_key.sign(&body),
        };
        self.issued.insert((slot, nullifier), ticket.digest()?);
        Ok(ticket)
    }
}

#[derive(Clone, Debug)]
pub struct RandomnessBeacon {
    pub round: u64,
    pub value: [u8; 32],
    pub signature: Signature,
}

impl RandomnessBeacon {
    fn unsigned(round: u64, value: &[u8; 32]) -> Vec<u8> {
        [BEACON_DOMAIN, &round.to_be_bytes(), value].concat()
    }

    pub fn sign(round: u64, value: [u8; 32], key: &SigningKey) -> Self {
        Self {
            round,
            value,
            signature: key.sign(&Self::unsigned(round, &value)),
        }
    }

    pub fn verify(&self, key: &VerifyingKey) -> bool {
        key.verify(&Self::unsigned(self.round, &self.value), &self.signature)
            .is_ok()
    }
}

#[derive(Clone, Debug)]
pub struct AdmissionReceipt {
    pub slot: u64,
    pub node: u64,
    pub ticket_digest: [u8; 32],
    pub frame_digest: [u8; 32],
    pub received_at_ns: u64,
    pub signature: Signature,
}

impl AdmissionReceipt {
    pub fn unsigned(&self) -> Vec<u8> {
        [
            RECEIPT_DOMAIN,
            &self.slot.to_be_bytes(),
            &self.node.to_be_bytes(),
            &self.ticket_digest,
            &self.frame_digest,
            &self.received_at_ns.to_be_bytes(),
        ]
        .concat()
    }

    pub fn verify(&self, key: &VerifyingKey) -> bool {
        key.verify(&self.unsigned(), &self.signature).is_ok()
    }
}

#[derive(Clone, Debug)]
pub struct BatchManifest {
    pub slot: u64,
    pub node: u64,
    pub beacon_round: u64,
    pub beacon_value: [u8; 32],
    pub ordered_ticket_digests: Vec<[u8; 32]>,
    pub ordered_frame_digests: Vec<[u8; 32]>,
    pub previous_digest: [u8; 32],
    pub signature: Signature,
}

impl BatchManifest {
    pub fn unsigned(&self) -> Result<Vec<u8>, String> {
        if self.ordered_ticket_digests.len() != self.ordered_frame_digests.len() {
            return Err("ticket and frame manifests have different lengths".into());
        }
        let mut body = Vec::new();
        body.extend_from_slice(MANIFEST_DOMAIN);
        body.extend_from_slice(&self.slot.to_be_bytes());
        body.extend_from_slice(&self.node.to_be_bytes());
        body.extend_from_slice(&self.beacon_round.to_be_bytes());
        body.extend_from_slice(&self.beacon_value);
        body.extend_from_slice(&self.previous_digest);
        body.extend_from_slice(&(self.ordered_ticket_digests.len() as u64).to_be_bytes());
        for (ticket, frame) in self
            .ordered_ticket_digests
            .iter()
            .zip(&self.ordered_frame_digests)
        {
            body.extend_from_slice(ticket);
            body.extend_from_slice(frame);
        }
        Ok(body)
    }

    pub fn digest(&self) -> Result<[u8; 32], String> {
        Ok(digest(&[&self.unsigned()?, &self.signature.to_bytes()]))
    }

    pub fn verify(&self, key: &VerifyingKey) -> bool {
        self.unsigned()
            .is_ok_and(|body| key.verify(&body, &self.signature).is_ok())
    }

    pub fn includes(&self, receipt: &AdmissionReceipt) -> bool {
        self.ordered_ticket_digests
            .iter()
            .zip(&self.ordered_frame_digests)
            .any(|(ticket, frame)| {
                ticket == &receipt.ticket_digest && frame == &receipt.frame_digest
            })
    }
}

pub struct FixedSlotSealer {
    pub slot: u64,
    pub node: u64,
    pub deadline_ns: u64,
    pub tickets: Vec<AdmissionTicket>,
    pub authority_key: VerifyingKey,
    pub beacon_key: VerifyingKey,
    pub signing_key: SigningKey,
    pub previous_digest: [u8; 32],
    frames: BTreeMap<[u8; 32], Frame>,
    receipts: BTreeMap<[u8; 32], AdmissionReceipt>,
    closed: bool,
}

impl FixedSlotSealer {
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        slot: u64,
        node: u64,
        deadline_ns: u64,
        tickets: Vec<AdmissionTicket>,
        authority_key: VerifyingKey,
        beacon_key: VerifyingKey,
        signing_key: SigningKey,
        previous_digest: [u8; 32],
    ) -> Result<Self, String> {
        let ids = tickets
            .iter()
            .map(|ticket| ticket.ticket_id)
            .collect::<BTreeSet<_>>();
        if ids.len() != tickets.len() {
            return Err("the expected ticket list contains a duplicate".into());
        }
        if tickets.iter().any(|ticket| ticket.slot != slot) {
            return Err("a ticket belongs to another slot".into());
        }
        Ok(Self {
            slot,
            node,
            deadline_ns,
            tickets,
            authority_key,
            beacon_key,
            signing_key,
            previous_digest,
            frames: BTreeMap::new(),
            receipts: BTreeMap::new(),
            closed: false,
        })
    }

    pub fn verifying_key(&self) -> VerifyingKey {
        self.signing_key.verifying_key()
    }

    pub fn admit(
        &mut self,
        ticket: &AdmissionTicket,
        frame: Frame,
        now_ns: u64,
    ) -> Result<AdmissionReceipt, String> {
        if self.closed {
            return Err("the slot is already closed".into());
        }
        if now_ns > self.deadline_ns {
            return Err("the frame arrived after the sealed deadline".into());
        }
        if !ticket.verify(&self.authority_key, now_ns / 1_000_000_000) {
            return Err("the admission ticket is invalid or expired".into());
        }
        let ticket_digest = ticket.digest()?;
        let expected = self
            .tickets
            .iter()
            .map(AdmissionTicket::digest)
            .collect::<Result<BTreeSet<_>, _>>()?;
        if !expected.contains(&ticket_digest) {
            return Err("the ticket was not in the slot's precommitted population".into());
        }
        if u64::from(frame.slot) != self.slot || u64::from(frame.node) != self.node {
            return Err("the frame belongs to another slot or node".into());
        }
        let raw = frame.encode();
        let frame_digest = Sha256::digest(raw).into();
        if let Some(prior) = self.frames.get(&ticket_digest) {
            if prior.encode() != raw {
                return Err("one ticket attempted to replace its admitted frame".into());
            }
            return Ok(self.receipts[&ticket_digest].clone());
        }
        let mut receipt = AdmissionReceipt {
            slot: self.slot,
            node: self.node,
            ticket_digest,
            frame_digest,
            received_at_ns: now_ns,
            signature: Signature::from_bytes(&[0; 64]),
        };
        receipt.signature = self.signing_key.sign(&receipt.unsigned());
        self.frames.insert(ticket_digest, frame);
        self.receipts.insert(ticket_digest, receipt.clone());
        Ok(receipt)
    }

    pub fn close(
        &mut self,
        beacon: &RandomnessBeacon,
        now_ns: u64,
    ) -> Result<(Vec<Frame>, BatchManifest), String> {
        if self.closed {
            return Err("the slot is already closed".into());
        }
        if now_ns <= self.deadline_ns {
            return Err("the batch cannot close before its deadline".into());
        }
        if !beacon.verify(&self.beacon_key) {
            return Err("the ordering beacon signature is invalid".into());
        }
        if beacon.round <= self.slot {
            return Err("ordering randomness must be generated after the slot".into());
        }
        let missing = self
            .tickets
            .iter()
            .filter(|ticket| {
                ticket
                    .digest()
                    .is_ok_and(|digest| !self.frames.contains_key(&digest))
            })
            .count();
        if missing != 0 {
            return Err(format!(
                "fixed population incomplete: {missing} cover or request frame(s) missing"
            ));
        }
        let mut ordered = self.tickets.clone();
        ordered.sort_by_key(|ticket| {
            Sha256::new()
                .chain_update(ORDER_DOMAIN)
                .chain_update(self.slot.to_be_bytes())
                .chain_update(beacon.round.to_be_bytes())
                .chain_update(beacon.value)
                .chain_update(ticket.ticket_id)
                .finalize()
                .to_vec()
        });
        let ticket_digests = ordered
            .iter()
            .map(AdmissionTicket::digest)
            .collect::<Result<Vec<_>, _>>()?;
        let frames = ticket_digests
            .iter()
            .map(|digest| self.frames[digest].clone())
            .collect::<Vec<_>>();
        let frame_digests = frames
            .iter()
            .map(|frame| Sha256::digest(frame.encode()).into())
            .collect();
        let mut manifest = BatchManifest {
            slot: self.slot,
            node: self.node,
            beacon_round: beacon.round,
            beacon_value: beacon.value,
            ordered_ticket_digests: ticket_digests,
            ordered_frame_digests: frame_digests,
            previous_digest: self.previous_digest,
            signature: Signature::from_bytes(&[0; 64]),
        };
        manifest.signature = self.signing_key.sign(&manifest.unsigned()?);
        self.closed = true;
        Ok((frames, manifest))
    }
}

pub fn prove_omission(
    receipt: &AdmissionReceipt,
    manifest: &BatchManifest,
    sealer_key: &VerifyingKey,
) -> bool {
    receipt.slot == manifest.slot
        && receipt.node == manifest.node
        && receipt.verify(sealer_key)
        && manifest.verify(sealer_key)
        && !manifest.includes(receipt)
}
